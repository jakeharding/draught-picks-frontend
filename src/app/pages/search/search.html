<!--
  Generated template for the SearchPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<!--<draught-header></draught-header>-->

<ion-content padding>
  <ion-row>
    <ion-col class="below-header" text-center col-12 col-sm-10 offset-sm-1 col-md-6 offset-md-3 col-lg-6 offset-lg-3 col-xl-4 offset-xl-4>
      <h1 text-left>What are you drinking?</h1>
      <ion-searchbar [(ngModel)]="beerSearch" (ionClear)="clear($event)" (ionInput)="search($event)" placeholder="Search Beer By Name"></ion-searchbar>
      <div *ngIf="beerResults && beerResults.length > 0">
        <div class="list-container" id="infinite-scroller" appInfiniteScroller [immediateCallback]="false" [scrollCallback]="scrollCallback" [loadMore]="loadMore">
          <beer-list [beers]="beerResults" ></beer-list>
        </div>
      </div>
      <div class="notYetSearch" text-left *ngIf="!beerResults || beerResults.length == 0" >
        <p><i>{{ message }} </i></p>
      </div>

    </ion-col>
  </ion-row>
</ion-content>

