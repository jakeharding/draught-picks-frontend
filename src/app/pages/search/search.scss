page-search {
.notYetSearch {
  padding-top: 1%;
  font-size: medium;
}
  .list-container {
    overflow-y: auto;
    position: relative;
    height: 600px;
  }
}
