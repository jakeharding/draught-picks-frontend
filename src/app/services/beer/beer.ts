import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment as Env } from '../../../environments/environment';
import Beer from '../../models/Beer';
import PageResponse from '../../models/PageResponse';
import { LIMIT } from '../../directives/infinite-scroller/infinite-scroller';
import { EMPTY, Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { AutoCompleteService } from 'ionic4-auto-complete';
/*
  Generated class for the BeerProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable({
  providedIn: 'root',
})
export class BeerProvider implements AutoCompleteService {
  url: string;
  recentsUrl: string;
  recommendedUrl: string;
  labelAttribute = 'name'; // Attribute needed for the AutoCompleteService
  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/beers`;
    this.recentsUrl = `${Env.REST_API_ROOT}/recent-beers`;
    this.recommendedUrl = `${Env.REST_API_ROOT}/recommended-beers`;
  }

  /**
   * search function
   * Parameters: Observable of type Beer list
   * Sets the limit and offset for the infinite scrolling
   */
  search(params: any): Observable<Beer[]> {
    if (!params.limit) {
      params.limit = LIMIT;
    }
    if (!params.offset) {
      params.offset = 0;
    }
    return this.http.get(this.url, {params}).pipe(map(({results}: PageResponse<Beer>) => results));
  }

  /**
   * Method called in auto complete search bar.
   * @param beerName - Name of beer
   * @returns - List of beers
   */
  getResults(beerName: string) {
    if (beerName) {
      return this.search({search: beerName});
    }
    return EMPTY;
  }

  /**
   * recents function
   * Parameters: Observable of type Beer list
   * Sets the limit and offset for the infinite scrolling
   */
  recents(params: any): Observable<Beer[]> {
    if (!params.limit) {
      params.limit = LIMIT;
    }
    if (!params.offset) {
      params.offset = 0;
    }

    return this.http.get(this.recentsUrl, {params}).pipe(map(({results}: PageResponse<Beer>) => results));
  }

  /**
   * recommended function
   * Parameters: query params for pagination
   * Retrieves a page of recommended beers for an authenticated user.
   */
  recommended(params: any): Observable<Beer[]> {
    if (!params.limit) {
      params.limit = LIMIT;
    }
    if (!params.offset) {
      params.offset = 0;
    }
    return this.http.get(this.recommendedUrl, {params})
      .pipe(map(({results}: PageResponse<Beer>) => results));
  }

  /**
   * createRecent function
   * Parameters: Beer
   * returns the recent beer list
   */
  createRecent(recent: Beer) {
    return this.http.post(this.recentsUrl, {beer: recent.uuid}).toPromise();
  }

  /**
   * Retrieve a information about a beer.
   * @param uuid - ID of beer
   * @returns - Promise a beer
   */
  retrieve(uuid: string) {
    return this.http.get(`${this.url}/${uuid}`).toPromise();
  }


}
