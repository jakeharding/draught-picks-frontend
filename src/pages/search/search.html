<!--
  Generated template for the SearchPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <ion-navbar color="primary">
    <ion-title text-center>Draught Picks</ion-title>
  </ion-navbar>
</ion-header>

<ion-content padding>
  <ion-row>
    <ion-col text-center col-12 col-sm-10 offset-sm-1 col-md-6 offset-md-3 col-lg-6 offset-lg-3 col-xl-4 offset-xl-4>


      <ion-searchbar [(ngModel)]="beerSearch" (ionInput)="search($event)" placeholder="Search Beer By Name"></ion-searchbar>
      <!--&lt;!&ndash;<ion-list>&ndash;&gt;-->
        <!--<ion-item *ngFor="let beer of beerResults">-->
          <!--{{ beer?.name }}-->
        <!--</ion-item>-->
      <!--&lt;!&ndash;</ion-list>&ndash;&gt;-->

      <beer-list [beers]="beerResults" ></beer-list>

    </ion-col>
  </ion-row>
</ion-content>

