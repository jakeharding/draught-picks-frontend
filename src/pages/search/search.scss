page-search {
.notYetSearch {
  padding-top: 1%;
  font-size: medium;
}
}
