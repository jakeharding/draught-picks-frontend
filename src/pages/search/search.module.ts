import { NgModule } from '@angular/core';
import { IonicPageModule } from 'ionic-angular';
import { SearchPage } from './search';
import { ComponentsModule } from "../../components/components.module";
import { DirectivesModule } from "../../directives/directives.module";

@NgModule({
  declarations: [
    SearchPage,
  ],
  imports: [
    IonicPageModule.forChild(SearchPage),
    ComponentsModule,
    DirectivesModule
  ],
})
export class SearchPageModule {}
