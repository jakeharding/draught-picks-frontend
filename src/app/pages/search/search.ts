import { Component } from '@angular/core';
import { NavController, NavParams } from '@ionic/angular';
import Beer from '../../models/Beer';
import { BeerProvider } from '../../services/beer/beer';
import { EMPTY } from 'rxjs';
import { LIMIT } from '../../directives/infinite-scroller/infinite-scroller';
import { BasePage } from '../BasePage';
import { tap } from 'rxjs/operators';

/**
 * Generated class for the SearchPage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@Component({
  selector: 'page-search',
  templateUrl: 'search.html',
  styleUrls: ['./search.scss']
})
export class SearchPage extends BasePage {
  private static NULL_RESULT_MESSAGE = 'Search for a beer and let us know what you like about it.';
  private static ZERO_RESULT_MESSAGE = 'We cannot find results on the beers you have entered.';

  beerResults: Array<Beer>;
  beerSearch: string;
  message: string;
  offset: number;
  loadMore: boolean;
  // scrollCallback;

  constructor(public navCtrl: NavController,
              public beerProvider: BeerProvider) {
    super('search');
    this.message = SearchPage.NULL_RESULT_MESSAGE;
    this.beerSearch = '';
    this.offset = 0;
    this.loadMore = true;
    // this.scrollCallback = this.getBeers.bind(this);
  }

  /**
   * Method called on input to the search bar.
   * @param event - Event object
   */
  search(event: Event) {
    if (this.beerSearch && this.beerSearch.length > 2) {
        this.beerProvider.search({search: this.beerSearch}).toPromise().then((results: Array<Beer>) => {
            this.beerResults = results;
            if (this.beerResults.length === 0) {
                this.message = SearchPage.ZERO_RESULT_MESSAGE;
            }
        });
    } else {
        this.beerResults = [];
        this.message = SearchPage.NULL_RESULT_MESSAGE;
    }
  }

  /**
   * getBeers function
   * No Parameters
   * gets the recent and recommended beers from the database and processes them
   * using the scrolling technique
   */
  getBeers(event) {
    if (this.loadMore) {
      const queryParams = {
        limit: LIMIT,
        offset: this.offset,
        search: this.beerSearch
      };
      return this.beerProvider.search(queryParams).pipe(tap(this.processData));
    }
    event.target.complete();
    return EMPTY;
  }

  /**
   * processData function
   * Parameters: Beer list
   * increases the offset and loads more beers while scrolling through the beer list
   */
  private processData = (beers) => {
    if (beers.length === 0) {
      this.loadMore = false;
      return;
    }
    this.offset += LIMIT;
    this.beerResults = this.beerResults.concat(beers);
  }

  /**
   * Method called when the `X` is clicked in the search bar.
   * @param event - event object
   */
  clear(event: Event): void {
    this.message = SearchPage.NULL_RESULT_MESSAGE;
    this.beerResults = null;

  }

  /**
   * Method called when view is about to enter.
   * Call the search method in case user has a value in the input.
   */
  ionViewWillEnter() {
    this.search(null);
  }
}
