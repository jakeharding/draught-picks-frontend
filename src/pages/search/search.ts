import { Component } from '@angular/core';
import {IonicPage, NavController, NavParams} from 'ionic-angular';
import { LoginRequired } from "../../providers/auth/auth";
import Beer from "../../models/Beer";
import {BeerProvider} from "../../providers/beer/beer";
import {Observable} from "rxjs/Observable";
import {LIMIT} from "../../directives/infinite-scroller/infinite-scroller";
import {BasePage} from "../BasePage";

/**
 * Generated class for the SearchPage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@LoginRequired
@IonicPage()
@Component({
  selector: 'page-search',
  templateUrl: 'search.html',
})
export class SearchPage extends BasePage {

  beerResults: Array<Beer>;
  beerSearch: string;
  message: string;
  offset: number;
  loadMore: boolean;
  scrollCallback;

  private static NULL_RESULT_MESSAGE = "Search for a beer and let us know what you like about it.";
  private static ZERO_RESULT_MESSAGE = "We cannot find results on the beers you have entered.";
  constructor(public navCtrl: NavController, public navParams: NavParams,
              public beerProvider: BeerProvider) {
    super('search');
    this.message = SearchPage.NULL_RESULT_MESSAGE;
    this.beerSearch = "";
    this.offset = 0;
    this.loadMore = true;
    this.scrollCallback = this.getBeers.bind(this);
  }

  /**
   * Method called on input to the search bar.
   * @param {Event} event
   */
  search (event:Event) {
    if(this.beerSearch && this.beerSearch.length > 2) {
        this.beerProvider.search({search: this.beerSearch}).toPromise().then((results: Array<Beer>) => {
            this.beerResults = results;
            if(this.beerResults.length === 0) {
                this.message = SearchPage.ZERO_RESULT_MESSAGE;
            }
        });
    } else {
        this.beerResults = [];
        this.message = SearchPage.NULL_RESULT_MESSAGE;
    }
  }

  /**
   * getBeers function
   * No Parameters
   * gets the recent and recommended beers from the database and processes them
   * using the scrolling technique
   * */
  getBeers(){
    if(this.loadMore){
      let queryParams = {
        limit: LIMIT,
        offset: this.offset,
        search: this.beerSearch
      };
      return this.beerProvider.search(queryParams).do(this.processData);
    }
    return Observable.empty();
  }

  /**
   * processData function
   * Parameters: Beer list
   * increases the offset and loads more beers while scrolling through the beer list
   * */
  private processData = (beers) => {
    if(beers.length == 0){
      this.loadMore = false;
      return;
    }
    this.offset += LIMIT;
    this.beerResults = this.beerResults.concat(beers);
  };

  /**
   * Method called when the `X` is clicked in the search bar.
   * @param {Event} event
   */
  clear(event:Event): void {
    this.message = SearchPage.NULL_RESULT_MESSAGE;
    this.beerResults = null;

  }

  /**
   * Method called when view is about to enter.
   * Call the search method in case user has a value in the input.
   */
  ionViewWillEnter() {
    this.search(null);
  }
}
