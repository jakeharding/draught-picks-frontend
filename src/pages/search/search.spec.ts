import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { SearchPage } from './search';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NavController, NavParams } from 'ionic-angular';
import { InfiniteScrollerDirective, LIMIT } from '../../directives/infinite-scroller/infinite-scroller';
import { BeerProvider } from '../../providers/beer/beer';
import { Observable } from 'rxjs';
import Beer from '../../models/Beer';
import mock = jest.mock;

/**
 * search.spec.ts
 *
 * Created by jake
 * Created on 2018-11-20
 * 
 * Test SearchPage.
 */

describe('SearchPage', () => {
  let fixture: ComponentFixture<SearchPage>;
  let component: SearchPage;

  const mockBeerProvider = {
    search: jest.fn(() => Observable.of([]))
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [SearchPage, InfiniteScrollerDirective],
      providers: [
        { provide: NavController, useValue: {}},
        { provide: NavParams, useValue: {}},
        { provide: BeerProvider, useValue: mockBeerProvider}
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SearchPage);
    component = fixture.debugElement.componentInstance;
  });

  it('should initialize the SearchPage', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  test('ionViewWillEnter will reset beerResults and message', () => {
    component.ionViewWillEnter();
    expect(component.beerResults).toEqual([]);
    expect(component.message).toEqual('Search for a beer and let us know what you like about it.');
  });

  test('search with call beerProvider.search to search for beers', async () => {
    const search = 'Bud';
    component.beerSearch = search;
    await component.search({} as Event);
    expect(mockBeerProvider.search).toHaveBeenCalledTimes(1);
    expect(mockBeerProvider.search).toHaveBeenCalledWith({search});
  });

  test('getBeers calls beerProvider.search when loadMore is true', async () => {
    component.loadMore = true;
    await component.getBeers();
    expect(mockBeerProvider.search).toHaveBeenCalledTimes(1);
  });

  test("getBeers doesn't calls beerProvider.search when loadMore is false", async () => {
    component.loadMore = false;
    const result = await component.getBeers();
    expect(mockBeerProvider.search).not.toHaveBeenCalled();
    expect(result).toEqual(Observable.empty());
  });

  test('processData sets loadMore to false when called with empty list', () => {
    component['processData']([]);
    expect(component.loadMore).toBe(false);
  });

  test('processData sets offset and adds to beerResults', () => {
    const mockBeer = {uuid:'beer'} as Beer;
    const currentOffset = component.offset;
    component.beerResults = [];
    component['processData']([mockBeer]);
    expect(component.loadMore).toBe(true);
    expect(component.offset).toBe(currentOffset + LIMIT);
    expect(component.beerResults).toContain(mockBeer);
  });

  it('should reset the message and beerResults', () => {
    component.clear({} as Event);
    expect(component.message).toEqual('Search for a beer and let us know what you like about it.');
    expect(component.beerResults).toBeNull();
  });
});
