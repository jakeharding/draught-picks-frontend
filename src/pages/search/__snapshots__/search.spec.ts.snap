// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage should initialize the SearchPage 1`] = `
<page-search
  beerProvider={[Function Object]}
  beerSearch=""
  loadMore={[Function Boolean]}
  message={[Function String]}
  navCtrl={[Function Object]}
  navParams={[Function Object]}
  offset="0"
  processData={[Function Function]}
  scrollCallback={[Function Function]}
>
  
  <draught-header />
  
  <ion-content
    padding=""
  >
    
    
    <ion-row>
      
      
      <ion-col
        class="below-header"
        col-12=""
        col-lg-6=""
        col-md-6=""
        col-sm-10=""
        col-xl-4=""
        offset-lg-3=""
        offset-md-3=""
        offset-sm-1=""
        offset-xl-4=""
        text-center=""
      >
        
        
        <h1
          text-left=""
        >
          What are you drinking?
        </h1>
        
        
        <ion-searchbar
          placeholder="Search Beer By Name"
        />
        
        
        
        
        
        
        
  
      
      </ion-col>
      
    
    </ion-row>
    
  
  </ion-content>
  
  
</page-search>
`;
